File: Difficulty: Hard/Alien Dictionary/alien-dictionary.cpp
//{ Driver Code Starts
// Initial Template for C++

#include <bits/stdc++.h>
using namespace std;

// } Driver Code Ends
// User function Template for C++

class Solution {
    
  void dictorder(string& str1, string& str2 ,pair<char,char>& ans) {
      int i = 0,j = 0;
      int asz = str1.length(), bsz = str2.length();
      
      ans.first = 'a' -1;
      ans.second = 'a' - 1;
      
      bool flag = false;
      while(i<asz && j<bsz) {
          if(str1[i]!=str2[j]) {
              flag = true;
              ans.first = str1[i];
              ans.second = str2[j];
              return;
              //break;
          }
          i++;j++;
      }
      if(i<asz) {
          ans.first = str1[i];
      }
      if(j<bsz) {
          ans.second = str2[j];
      }
  }
  public:
    string findOrder(string dict[], int n, int k) {
        // code here
        vector<vector<int> > graph(k);
        
        
        
        for(int i = 1;i<n;i++) {
            
            pair<char,char> order;
            dictorder(dict[i-1],dict[i],order);
            
            if(order.first>='a' && order.second>='a')
                graph[order.first-'a'].push_back(order.second-'a');
            
            
            
        }
        
        
        vector<int> indeg(k);
        
        for(int i = 0;i<k;i++) {
            for(auto a:graph[i]) {
                indeg[a]++;
            }
        }
        queue<int> que;
        for(int i = 0;i<k;i++) {
            if(indeg[i]==0) que.push(i);
        }
        string ans;
        while(!que.empty()) {
            int node = que.front();
            que.pop();
            ans+=(node+'a');
            for(auto a:graph[node]) {
                indeg[a]--;
                if(indeg[a]==0) que.push(a);
            }
        }
        return ans;
    }
};

//{ Driver Code Starts.
string order;

bool f(string a, string b) {
    int p1 = 0;
    int p2 = 0;
    for (int i = 0; i < min(a.size(), b.size()) and p1 == p2; i++) {
        p1 = order.find(a[i]);
        p2 = order.find(b[i]);
        //	cout<<p1<<" "<<p2<<endl;
    }

    if (p1 == p2 and a.size() != b.size())
        return a.size() < b.size();

    return p1 < p2;
}

// Driver program to test above functions
int main() {
    int t;
    cin >> t;
    while (t--) {
        int N, K;
        cin >> N >> K;
        string dict[N];
        for (int i = 0; i < N; i++)
            cin >> dict[i];

        Solution obj;
        string ans = obj.findOrder(dict, N, K);
        order = "";
        for (int i = 0; i < ans.size(); i++)
            order += ans[i];

        string temp[N];
        std::copy(dict, dict + N, temp);
        sort(temp, temp + N, f);

        bool f = true;
        for (int i = 0; i < N; i++)
            if (dict[i] != temp[i])
                f = false;

        if (f)
            cout << 1;
        else
            cout << 0;
        cout << endl;
    }
    return 0;
}

// } Driver Code Ends

File: Difficulty: Hard/Alien Dictionary/alien-dictionary.java
class Solution {
	
	
	public void constructAdjlist(String[] words, ArrayList<Set<Integer>> adjlist) {
		
		int n = words.length;
		for (int i = 1; i<n; i++) {
			String s1 = words[i - 1];
			
			String s2 = words[i];
			
			int m = Math.min(s1.length(), s2.length());
			
			for (int k = 0; k<m; k++) {
				if (s1.charAt(k) != s2.charAt(k)) {
					int u = s1.charAt(k) - 'a';
					int v = s2.charAt(k) - 'a';
					// System.out.print(u+" -> "+v+", ");
					adjlist.get(u).add(v);
					break;
				}
			}
			
			
		}
		
	}
	
	public String findOrder(String[] words) {
		// code here
		ArrayList<Set<Integer>> adjlist = new ArrayList<>();
		for (int i = 0; i<26; i++)
			adjlist.add(new HashSet<Integer>());
		
		int n = words.length;
		int[] indeg = new int[26];
		
		Arrays.fill(indeg, -1);
		
		for (int i = 0; i<n; i++) {
			int m = words[i].length();
			for (int j = 0; j<m; j++) {
				indeg[words[i].charAt(j) - 'a'] = 0;
			}
		}
		
		constructAdjlist(words, adjlist);
		
		for (int u = 0; u<26; u++) {
			
			for (Integer v: adjlist.get(u)) {
				indeg[v] = (indeg[v] == -1)?1:indeg[v]+1;
			}
		}
		
		Queue<Integer> que = new ArrayDeque<>();
		
		int total = 0;
		
		for (int i = 0; i<26; i++) {
			if (indeg[i] == 0) {
				que.add(i);
			}
			if (indeg[i] >= 0)
				total++;
		}
		
		/// int ans = 0;
		
		StringBuilder ans = new StringBuilder();
		
		while (!que.isEmpty()) {
			int u = que.poll();
			ans.append((char)('a' + u));
			for (Integer v: adjlist.get(u)) {
				indeg[v]--;
				if (indeg[v] == 0) {
					que.add(v);
				}
			}
		}
		
		// System.out.println(ans+", "+total);
		
		if (ans.length() != total)
			return "";
		
		// return "true";
		
		// System.out.println(ans.toString());
		return (ans.toString());
	}
}
